### DISCUSSION TITLE
Fastest 2D video mode

### DISCUSSION THREAD

**Discussion by rofl0r** [pspdev/pspsdk#187] [Q&A] [answered]
currently using SDL2 hw-accelerated 32bit RGBA mode, i'm getting a mere 16 fps. which videomode works best on the GPU in the PSP ?

**Comment by sharkwouter** [comment]
How are you use it? SDL2 is okay for drawing shapes and sprites, but for generating and displaying surfaces pixel by pixel the implementation is slow. You'll probably have better luck using libgu or libgum directly. There are some examples available here: https://github.com/pspdev/pspsdk/tree/master/src/samples/gu

**Reply by rofl0r** [reply]
yeah, i use a Surface/Texture that basically represents the full screen, using it in framebuffer like fashion. assuming that at least some games use SDL2 though, i suspect there must be a way to use it and to get a decent performance without punching in raw GL/GU commands. thanks for the link.

**Reply by sharkwouter** [reply]
I think SDL needs some optimization, but I've not been able to figure out how to do that yet.

**Reply by rofl0r** [reply, marked as answer]
alright, after reading the source code of the PSP implementation it becomes clear that the PSP implements only 2 videomodes "natively": `SDL_PIXELFORMAT_ABGR8888` and `SDL_PIXELFORMAT_BGR565`. by natively i mean that these map directly to modes supported by the GU. using anything else basically requires SDL to recalculate every single pixel in the texture to one of those 2 formats.

i also found the function `SDL_LockTextureToSurface` which provides a temporary surface which let's me write directly to the texture's memory (which i assume is in VRAM).

using this function, the ABGR mode, and removing SDL_RenderClear() from the draw loop now give me 75 fps with nothing else changed.

the code now looks more or less like this:

```c
        /* setup */
        c->window = SDL_CreateWindow("sdl-console",
                             SDL_WINDOWPOS_UNDEFINED,
                             SDL_WINDOWPOS_UNDEFINED,
                             480, 272,
                             SDL_WINDOW_OPENGL);
        c->renderer = SDL_CreateRenderer(c->window, -1,
                SDL_RENDERER_ACCELERATED|SDL_RENDERER_PRESENTVSYNC|
                SDL_RENDERER_TARGETTEXTURE);
        c->texture = SDL_CreateTexture(c->renderer,
                SDL_PIXELFORMAT_ABGR8888, SDL_TEXTUREACCESS_STREAMING, 480, 272);

        /* draw loop */
        while(on) {
        SDL_Surface *out;
        SDL_LockTextureToSurface(c->texture, NULL, &out);
        /* ... manipulate out->pixels ... */
        SDL_UnlockTexture(c->texture);
        SDL_RenderCopy(c->renderer, c->texture, NULL, NULL);
        SDL_RenderPresent(c->renderer);
        }
```
        
75 fps is good enough for the moment, though i suspect by using BGR565 it will get even faster probably 100ish fps.
so i guess this can be marked as answered.

**Reply by sharkwouter** [reply]
Interesting, I'll have to try this to see if it improves performance for a project I'm working on.

**Reply by rofl0r** [reply]
small update:
1) the reported 75 fps were on PPSSPP. on real hardware more like 26-35 fps. it's fast enough though that the rendering doesn't interrupt the audio playback and the UI appears responsive.
2) bgr565 is indeed faster. the real fps on the device are 40-45, depending on whether audio is on.
3) requesting a texture/surface of 512 pixels width doesn't seem to make a difference (this is recommended in several raw GU tutorials). i could imagine SDL2 already rounds up the texture size internally.

all measurements were taken with 222mhz and vsync enabled, and redrawing the entire texture for each draw cycle. to achieve higher performance, one probably has to punch in raw GL/GU commands and/or use one texture per "sprite".
